### PR TITLE
Add XSS context analyzer with context-aware verification

### PR DESCRIPTION
**Pull request by orbin123**
# Context-Aware XSS Analyzer for Fuzzing Engine

**Issue:** #5838  

This PR introduces a **context-aware XSS analyzer** for the fuzzing engine. It detects reflection contexts, selects payloads based on context and available characters, and verifies exploitability to reduce false positives.

---

### 🔍 Detects:
- HTML body  
- Quoted and unquoted attributes  
- Script blocks, strings, and template literals  
- Style blocks  
- Comments  
- URL attributes  

### 🎯 Context-aware payload selection with:
- Character filtering  
- Configurable maximum attempts  
- Custom canary support via `[XSS_CANARY]` placeholder  

### ✅ Verification logic to:
- Reject non-exploitable reflections  
- Handle comments safely  
- Detect encoding issues  
- Identify context mismatches  

---

#### `pkg/fuzz/analyzers/xss/analyzer.go`
Orchestrates probe, detection, payload selection, and verification. Implements the Analyzer interface with:
- `ApplyInitialTransformation()` — Handles `[XSS_CANARY]` placeholder replacement  
- `Analyze()` — Main flow: send canary → detect contexts → select payloads → verify  
- `sendRequest()` — HTTP request handling with proper error handling for `resp.Body.Close()`  
- `verifyXSS()` — Confirms exploitability by checking context compatibility and encoding  

#### `pkg/fuzz/analyzers/xss/context_detector.go`
- Context detection heuristics  
- Character availability logic (well documented)

#### `pkg/fuzz/analyzers/xss/payload_selector.go`
- Context-specific payload sets  
- Filtering rules  

#### `pkg/fuzz/analyzers/xss/types.go`
Includes context/type definitions and analyzer constants:
- `ContextType` enum with 9 context types  
- `ReflectionInfo` struct with position, context, available chars  
- `CharacterSet` for tracking `< > ' " / \` availability  
- `DefaultCanary = "xSs9K7j<>'\"/()"`  
- `AnalyzerName = "xss_context"`

---

## Integration Tests

**integration_tests/fuzz/fuzz-xss-context-body.yaml**  
Tests XSS detection in HTML body context where input reflects inside `<div>` tags without encoding.

**integration_tests/fuzz/fuzz-xss-context-attribute.yaml**  
Tests XSS detection in HTML attribute context where input reflects inside `value=""` attributes.

**integration_tests/fuzz/fuzz-xss-context-script.yaml**  
Tests XSS detection in JavaScript context where input reflects inside `<script>var x = "..."</script>`.

**integration_tests/fuzz/fuzz-xss-context-encoded.yaml**  
Negative test case — verifies analyzer correctly identifies when XSS is **not exploitable** due to HTML encoding.

**cmd/integration-test/fuzz.go**  
Added 4 test handlers that create vulnerable/safe test servers:
- `xssContextBodyFuzz` — Vulnerable HTML body reflection  
- `xssContextAttributeFuzz` — Vulnerable attribute reflection  
- `xssContextScriptFuzz` — Vulnerable script context reflection  
- `xssContextEncodedFuzz` — Safe (encoded) reflection for negative testing  

---

## Documentation

**pkg/templates/templates_doc.go**
- Added `xss_context` to valid analyzer names list  
- Updated parameters description to mention the canary parameter  

---

### `pkg/fuzz/analyzers/xss/analyzer_test.go`
Comprehensive coverage including:
- Body, attribute, script, and URL context detection  
- No-reflection handling  
- Encoded reflection handling  
- Custom canary support  
- False positive rejection in comments  

### `pkg/fuzz/analyzers/xss/context_detector_test.go`
Covers:
- HTML body, attribute (quoted/unquoted), script, style, comment contexts  
- Multiple reflections handling  
- Case-insensitive tag detection  
- URL attribute detection (`href, src, action, data, formaction, poster`)  
- Script inside SVG handling  
- CDATA section handling  
- Partial encoding scenarios  

### `pkg/fuzz/analyzers/xss/payload_selector_test.go`
Tests:
- Context-specific payload selection  
- Character filtering  
- Max verification attempts  

---

## Edge Cases Covered
- Nested script contexts inside SVG  
- CDATA sections inside scripts  
- Partial encoding:
  - Only `<` encoded  
  - Only `>` encoded  
  - Quotes-only encoded  
  - Mixed encoding cases  
- HTML comments (non-exploitable)  
- Context compatibility verification  
- Critical character encoding detection  

---

## Usage Example

```yaml
id: xss-context-detection
info:
  name: XSS Context Analyzer
  author: pdteam
  severity: high
http:
  method: GET
  path:
    - "{{BaseURL}}"
fuzzing:
  part: query
  type: replace
  mode: single
  fuzz:
    - "[XSS_CANARY]"
analyzer:
  name: xss_context
  parameters:
    canary: "xSs9K7j<>'\"/()"
```

/claim #5838

## Summary by CodeRabbit

* **New Features**
  * Added a context-aware XSS analyzer with configurable canary, automated context detection, payload selection and verification across HTML, JS and URL contexts.

* **Tests**
  * Added extensive unit and integration tests covering context detection, payload selection/filtering, verification logic, encoding and edge cases; added fuzz scenarios.

* **Documentation**
  * Updated analyzer templates/docs to include the new XSS analyzer and customizable canary option.

* **Chores**
  * Registered the analyzer in the HTTP workflow and applied minor test/formatting adjustments.

### CODE REVIEW

File: pkg/fuzz/analyzers/xss/analyzer.go [hunk 1]
```diff
@@ -0,0 +1,252 @@
...
+
+	// Read response body (limit to 1MB to avoid memory issues)
+	limitReader := io.LimitReader(resp.Body, 1*1024*1024)
+	body, err := io.ReadAll(limitReader)
+	if err != nil {
+		return "", err
+	}
+
+	return string(body), nil
+}
```
[the remark this answers is not in the document]
Comment (orbin123) [reply pkg/fuzz/analyzers/xss/analyzer.go:132]: This pattern matches the existing time_delay analyzer implementation (lines 114-133 in pkg/fuzz/analyzers/time/analyzer.go). The Component interface is designed to be stateful and reused across multiple requests via SetValue() + Rebuild(). This is the established pattern in the analyzer framework and does not cause state leakage.

File: pkg/fuzz/analyzers/xss/analyzer.go [hunk 2]
```diff
@@ -0,0 +1,252 @@
...
+		payload,
+		originalReflection.AvailableChars.LessThan,
+		originalReflection.AvailableChars.GreaterThan,
+		originalReflection.AvailableChars.SingleQuote,
+		originalReflection.AvailableChars.DoubleQuote,
+		originalReflection.AvailableChars.Slash,
+	)
+
+	return true, details
+}
```
[the remark this answers is not in the document]
Comment (orbin123) [reply pkg/fuzz/analyzers/xss/analyzer.go:190]: The analyzer already iterates through all detected reflections in the Analyze() function (lines 78-94). Each reflection gets its own payload verification attempts. Finding one confirmed XSS is sufficient - we don't need to verify all occurrences of the same payload.

File: pkg/fuzz/analyzers/xss/context_detector.go [hunk 3]
```diff
@@ -0,0 +1,328 @@
...
+	return ReflectionInfo{
+		Position:       canaryPos,
+		Context:        context,
+		AvailableChars: chars,
+		BeforeCanary:   beforeCanary,
+		AfterCanary:    afterCanary,
+		AttributeName:  attributeName,
+		QuoteChar:      quoteChar,
+	}
+}
```
[the remark this answers is not in the document]
Comment (orbin123) [reply pkg/fuzz/analyzers/xss/context_detector.go]: This is intentional. If the entire canary is encoded, there's no exploitable reflection. The detection phase finds reflection points, then detectAvailableCharacters determines which specific characters survived encoding. This two-phase approach (detect → analyze available chars → select payloads) is the core design.

File: pkg/fuzz/analyzers/xss/context_detector.go [hunk 4]
```diff
@@ -0,0 +1,328 @@
...
+		quoteChar := getAttributeQuoteChar(lookback)
+		if quoteChar == "\"" || quoteChar == "'" {
+			return ContextHTMLAttributeQuoted
+		}
+		return ContextHTMLAttributeUnquoted
+	}
+
+	// Default to HTML body
+	return ContextHTMLBody
+}
```
[the remark this answers is not in the document]
Comment (orbin123) [reply pkg/fuzz/analyzers/xss/context_detector.go:147]: Valid point - will fix to properly detect quoted/URL attributes inside script/style opening tags.

**Comment by dogancanbakir** [comment]
@orbin123 Thanks for the PR, please make sure to add working examples to demonstrate how this feature is working. This is very important.

**Comment by orbin123** [comment]
> @orbin123 Thanks for the PR, please make sure to add working examples to demonstrate how this feature is working. This is very important.

Yes, I've added working examples demonstrating the feature: 4 integration test templates (fuzz-xss-context-body.yaml, fuzz-xss-context-attribute.yaml, fuzz-xss-context-script.yaml, fuzz-xss-context-encoded.yaml) that test XSS detection across HTML body, attribute, and script contexts, plus a negative test for encoded reflections. Each template has a corresponding test handler in cmd/integration-test/fuzz.go that creates a vulnerable test server to validate the analyzer's behavior.

**Comment by dogancanbakir** [comment]
@orbin123 This is the PR template, https://github.com/projectdiscovery/nuclei/blob/dev/.github/PULL_REQUEST_TEMPLATE.md

**Comment by orbin123** [comment]
## Proof

### Build & Test Output

https://github.com/user-attachments/assets/59e6b9c7-3792-4c25-a9f1-0b2776282460

### Real-World Demo Against Vulnerable Target
Two small local servers - One Vulnerable and another one safe

https://github.com/user-attachments/assets/c12eeaf7-f016-4e51-a5e3-d4c79f7633b7




### Integration Test Results
Screenshot 2026-02-03 at 6 16 09 PM


### Documentations

| Layer | Where | What it documents |
|------|-------|-------------------|
| **Package** | Top of `types.go` | One block: purpose, 3 phases (context → payloads → verification), scope/limits |
| **Types** | Same file | One-line comment per type; constants get inline examples (e.g. `// <div>CANARY</div>`) |
| **Functions** | Above each func in `analyzer.go`, `context_detector.go`, etc. | 1–3 lines: what it does, main inputs/return, limits (e.g. “up to 10 reflections”, “1MB body”) |
| **HTTP/template schema** | `http.go` + `templates_doc.go` | Analyzer field description; allowed names (`time_delay`, `xss_context`); parameters (e.g. `canary` for `xss_context`) |
| **Integration YAML** | `fuzz-xss-context-*.yaml` | `info.name` + `info.description`: which context is tested and that the analyzer picks payloads |
| **In-code** | Inside functions | Short step comments (“Step 1: Send canary”), guards (“limit to 10”), and non-obvious checks (“inside HTML comment”) |

File: pkg/fuzz/analyzers/xss/analyzer.go [hunk 5]
```diff
@@ -0,0 +1,280 @@
...
+
+// isContextCompatible checks if the verification context is compatible with the original context.
+// Exact matches are always compatible. Some contexts can evolve during exploitation:
+//   - Attribute contexts (quoted/unquoted/URL) can break out to HTML body
+//   - Script string context can break out to script block
+//
+// Returns true if contexts are compatible for XSS verification.
+func isContextCompatible(verifyContext, originalContext ContextType) bool {
+	// Exact match is always compatible
```
Comment (knakul853) [inline pkg/fuzz/analyzers/xss/analyzer.go]: for this method its missing script string→block tests, can you cover those, otherwise good test coverage.

File: pkg/fuzz/analyzers/xss/context_detector.go [hunk 6]
```diff
@@ -0,0 +1,374 @@
...
+		afterScript := lookbackLower[lastIndex:]
+		// Make sure we're inside the script tag, not after </script>
+		if !strings.Contains(afterScript, "</script>") {
+			// Check if the opening tag is properly closed with >
+			// This handles cases like <script src="CANARY"> where we're in an attribute
+			closingBracketPos := strings.Index(afterScript, ">")
+			if closingBracketPos == -1 {
+				// No closing bracket found, we're still in the opening tag attributes
+				return ContextHTMLAttributeUnquoted
+			}
```
Comment (knakul853) [inline pkg/fuzz/analyzers/xss/context_detector.go:101]:
```suggestion
			if closingBracketPos == -1 {
				
	       if isInAttributeContext(lookback) {
          quoteChar := getAttributeQuoteChar(lookback)
          if quoteChar == "\"" || quoteChar == "'" {
              return ContextHTMLAttributeQuoted
          }
      }
      return ContextHTMLAttributeUnquoted
			}
```

File: pkg/fuzz/analyzers/xss/context_detector.go [hunk 7]
```diff
@@ -0,0 +1,374 @@
...
+	// Check for style context
+	if lastIndex := strings.LastIndex(lookbackLower, "<style"); lastIndex != -1 {
+		afterStyle := lookbackLower[lastIndex:]
+		if !strings.Contains(afterStyle, "</style>") {
+			// Check if the opening tag is properly closed with >
+			closingBracketPos := strings.Index(afterStyle, ">")
+			if closingBracketPos == -1 {
+				// No closing bracket found, we're still in the opening tag attributes
+				return ContextHTMLAttributeUnquoted
+			}
```
Comment (knakul853) [inline pkg/fuzz/analyzers/xss/context_detector.go:123]:
```suggestion
			if closingBracketPos == -1 {
				if isInAttributeContext(lookback) {
          quoteChar := getAttributeQuoteChar(lookback)
          if quoteChar == "\"" || quoteChar == "'" {
              return ContextHTMLAttributeQuoted
          }
      }			}
```

File: pkg/fuzz/analyzers/xss/context_detector.go [hunk 8]
```diff
@@ -0,0 +1,374 @@
...
+// isInURLAttribute checks if the position is in a URL-bearing attribute such as
+// href, src, action, data, formaction, or poster. These attributes have special
+// XSS exploitation considerations (e.g., javascript: protocol). Returns true
+// if inside a URL attribute context.
+func isInURLAttribute(text string) bool {
+	// List of URL attributes (comprehensive list for XSS detection)
+	urlAttributes := []string{
+		"href", "src", "action", "data", "formaction", "poster",
+		"codebase", "cite", "background", "dynsrc", "lowsrc", "manifest",
+	}
```
Comment (knakul853) [inline pkg/fuzz/analyzers/xss/context_detector.go]: could be moved to types.go

File: pkg/fuzz/analyzers/xss/context_detector.go [hunk 9]
```diff
@@ -0,0 +1,374 @@
...
+		context != ContextHTMLAttributeUnquoted &&
+		context != ContextURLAttribute {
+		return "", ""
+	}
+
+	// Extract attribute name (word before =)
+	parts := strings.Split(beforeCanary, "=")
+	if len(parts) < 2 {
+		return "", ""
+	}
```
Comment (knakul853) [inline pkg/fuzz/analyzers/xss/context_detector.go]:
```suggestion
  lastEquals := strings.LastIndex(beforeCanary, "=")
  if lastEquals == -1 {
      return "", ""
  }
```

File: pkg/fuzz/analyzers/xss/context_detector.go [hunk 10]
```diff
@@ -0,0 +1,374 @@
...
+
+	attrPart := strings.TrimSpace(parts[len(parts)-2])
+	words := strings.Fields(attrPart)
+	if len(words) > 0 {
+		attrName := words[len(words)-1]
+		quoteChar := getAttributeQuoteChar(beforeCanary)
+		return attrName, quoteChar
+	}
```
Comment (knakul853) [inline pkg/fuzz/analyzers/xss/context_detector.go]:
```suggestion
 beforeEquals := strings.TrimSpace(beforeCanary[:lastEquals])
  words := strings.Fields(beforeEquals)
  if len(words) > 0 {
      attrName := words[len(words)-1]
      quoteChar := getAttributeQuoteChar(beforeCanary)
      return attrName, quoteChar
  }
```

File: pkg/fuzz/analyzers/xss/payload_selector.go [hunk 11]
```diff
@@ -0,0 +1,148 @@
...
+
+	// Filter by available characters
+	filtered := filterByAvailableChars(payloads, reflection.AvailableChars, reflection.Context)
+
+	// Limit number of attempts
+	maxAttempts := 3
+	if max, ok := params["max_verification_attempts"].(int); ok && max > 0 {
+		maxAttempts = max
+	}
+
```
Comment (knakul853) [inline pkg/fuzz/analyzers/xss/payload_selector.go:88]:
```suggestion
  maxAttempts := 3
  switch v := params["max_verification_attempts"].(type) {
  case int:
      if v > 0 {
          maxAttempts = v
      }
  case float64:
      if v > 0 {
          maxAttempts = int(v)
      }
  }

```

File: pkg/fuzz/analyzers/xss/analyzer.go [hunk 12]
```diff
@@ -0,0 +1,280 @@
...
+	// which could contain HTML tags that confuse context detection
+
+	// Find where the payload is reflected
+	payloadPos := strings.Index(responseBody, payload)
+	if payloadPos == -1 {
+		return false, ""
+	}
+
+	// Get lookback text to determine context at payload position
+	lookbackStart := max(0, payloadPos-500)
```
Comment (knakul853) [inline pkg/fuzz/analyzers/xss/analyzer.go]:
```suggestion
	lookbackStart := max(0, payloadPos-500)
```

here numbers are hardcoded  could cause mismatch, lets move these to types.go 
  const (
      DefaultCanary        = "xSs9K7j<>'\"/()"
      AnalyzerName         = "xss_context"
      contextLookbackSize  = 500
      surroundingTextSize  = 200
  )

File: pkg/fuzz/analyzers/xss/types.go [hunk 13]
```diff
@@ -0,0 +1,103 @@
...
+
+// CharacterSet tracks which special characters survived encoding
+type CharacterSet struct {
+	LessThan    bool // <
+	GreaterThan bool // >
+	SingleQuote bool // '
+	DoubleQuote bool // "
+	Slash       bool // /
+	Backtick    bool // `
+}
```
Comment (knakul853) [inline pkg/fuzz/analyzers/xss/types.go:96]:
```suggestion
  type CharacterSet struct {
      HasLessThan    bool // <
      HasGreaterThan bool // >
      HasSingleQuote bool // '
      HasDoubleQuote bool // "
      HasSlash       bool // /
      HasBacktick    bool // `
  }
```
in the context its used, this is much easier to read

**Review by knakul853 — CHANGES_REQUESTED** [review]

File: pkg/fuzz/analyzers/xss/analyzer.go [hunk 14]
```diff
@@ -0,0 +1,280 @@
...
+	// Step 2: Detect contexts
+	reflections := DetectContexts(responseBody, canary)
+	if len(reflections) == 0 {
+		gologger.Verbose().Msgf("[%s] No reflections found", a.Name())
+		return false, "", nil // No reflection found
+	}
+
+	gologger.Verbose().Msgf("[%s] Found %d reflection(s)", a.Name(), len(reflections))
+
+	// Step 3: Try verification for each reflection
```
Comment (knakul853) [inline pkg/fuzz/analyzers/xss/analyzer.go]: there are lots of inline comments which not needed, please have a look our calude.md file about comment rules

**Review by knakul853 — APPROVED** [review]

**Comment by orbin123** [comment]
Requesting @dwisiswant0 @Ice3man543 @Mzack9999 for further review.

**Review by Mzack9999 — CHANGES_REQUESTED** [review]
- [ ] Failing tests

**Comment by Mzack9999** [comment]
Closing this PR as Issue #5838 has been resolved via PR #7164, which implements the XSS context analyzer with HTML5-compliant tokenization and proper framework integration. While this PR includes valuable additions like integration tests and payload selection, the merged solution takes a different architectural approach with a more focused scope. Thank you for the comprehensive contribution.
